The bugfix may require some extra patching
```
tensorflow/core/distributed_runtime/tensor_coding.cc:85:39: error: too many arguments to function call, expected single argument 'total_bytes_limit', have 2 arguments
    input.SetTotalBytesLimit(INT_MAX, INT_MAX);  // Unlimited
    ~~~~~~~~~~~~~~~~~~~~~~~~          ^~~~~~~
/nix/store/gi4birnm1fb1ndq04qhzk5zam23za82j-clang-wrapper-7.1.0/resource-root/include/limits.h:62:19: note: expanded from macro 'INT_MAX'
#define INT_MAX   __INT_MAX__
                  ^~~~~~~~~~~
<built-in>:76:21: note: expanded from here
#define __INT_MAX__ 2147483647
                    ^~~~~~~~~~
/nix/store/209sqr8p9xivgwdgmxycipns5vy9p0ys-protobuf-3.19.0/include/google/protobuf/io/coded_stream.h:406:3: note: 'SetTotalBytesLimit' declared here
  void SetTotalBytesLimit(int total_bytes_limit);
  ^
tensorflow/core/distributed_runtime/tensor_coding.cc:221:37: error: too many arguments to function call, expected single argument 'total_bytes_limit', have 2 arguments
  input.SetTotalBytesLimit(INT_MAX, INT_MAX);  // Unlimited
  ~~~~~~~~~~~~~~~~~~~~~~~~          ^~~~~~~
/nix/store/gi4birnm1fb1ndq04qhzk5zam23za82j-clang-wrapper-7.1.0/resource-root/include/limits.h:62:19: note: expanded from macro 'INT_MAX'
#define INT_MAX   __INT_MAX__
                  ^~~~~~~~~~~
<built-in>:76:21: note: expanded from here
#define __INT_MAX__ 2147483647
                    ^~~~~~~~~~
/nix/store/209sqr8p9xivgwdgmxycipns5vy9p0ys-protobuf-3.19.0/include/google/protobuf/io/coded_stream.h:406:3: note: 'SetTotalBytesLimit' declared here
  void SetTotalBytesLimit(int total_bytes_limit);
  ^
2 errors generated.
```

--- a/pkgs/development/python-modules/tensorflow/default.nix
+++ b/pkgs/development/python-modules/tensorflow/default.nix
@@ -6,13 +6,13 @@
 , numpy, tensorflow-tensorboard_2, absl-py
 , setuptools, wheel, keras-preprocessing, google-pasta
 , opt-einsum, astunparse, h5py
-, termcolor, grpcio, six, wrapt, protobuf, tensorflow-estimator_2
+, termcolor, grpcio, six, wrapt, protobuf-python, tensorflow-estimator_2
 , dill, flatbuffers-python, tblib, typing-extensions
 # Common deps
 , git, pybind11, which, binutils, glibcLocales, cython, perl
 # Common libraries
 , jemalloc, mpi, gast, grpc, sqlite, boringssl, jsoncpp
-, curl, snappy, flatbuffers-core, lmdb-core, icu, double-conversion, libpng, libjpeg_turbo, giflib
+, curl, snappy, flatbuffers-core, lmdb-core, icu, double-conversion, libpng, libjpeg_turbo, giflib, protobuf-core
 # Upsteam by default includes cuda support since tensorflow 1.15. We could do
 # that in nix as well. It would make some things easier and less confusing, but
 # it would also make the default tensorflow package unfree. See
@@ -90,7 +90,7 @@ let
       keras-preprocessing
       numpy
       opt-einsum
-      protobuf
+      protobuf-python
       setuptools
       six
       tblib
@@ -124,13 +124,15 @@ let
       ./relax-dependencies.patch
       # Add missing `io_bazel_rules_docker` dependency.
       ./workspace.patch
+      # Patch the sources to compile with protobuf >= 3.16.
+      ./system-protobuf.patch
     ];
 
     # On update, it can be useful to steal the changes from gentoo
     # https://gitweb.gentoo.org/repo/gentoo.git/tree/sci-libs/tensorflow
 
     nativeBuildInputs = [
-      which pythonEnv cython perl
+      which pythonEnv cython perl protobuf-core
     ] ++ lib.optional cudaSupport addOpenGLRunpath;
 
     buildInputs = [
@@ -179,12 +181,7 @@ let
       # "com_github_googleapis_googleapis"
       # "com_github_googlecloudplatform_google_cloud_cpp"
       "com_github_grpc_grpc"
-      # Multiple issues with custom protobuf.
-      # First `com_github_googleapis` fails to configure. Can be worked around by disabling `com_github_googleapis`
-      # and related functionality, but then the next error is about "dangling symbolic link", and in general
-      # looks like that's only the beginning: see
-      # https://stackoverflow.com/questions/55578884/how-to-build-tensorflow-1-13-1-with-custom-protobuf
-      # "com_google_protobuf"
+      "com_google_protobuf"
       # Fails with the error: external/org_tensorflow/tensorflow/core/profiler/utils/tf_op_utils.cc:46:49: error: no matching function for call to 're2::RE2::FullMatch(absl::lts_2020_02_25::string_view&, re2::RE2&)'
       # "com_googlesource_code_re2"
       "curl"
@@ -219,6 +216,11 @@ let
     ];
 
     INCLUDEDIR = "${includes_joined}/include";
+
+    # This is needed for the Nix-provided protobuf dependency to work,
+    # as otherwise the rule `link_proto_files` tries to create the links
+    # to `/usr/include/...` which results in build failures.
+    PROTOBUF_INCLUDE_PATH = "${protobuf-core}/include";
 
     PYTHON_BIN_PATH = pythonEnv.interpreter;
 
@@ -295,12 +297,12 @@ let
     fetchAttrs = {
       # cudaSupport causes fetch of ncclArchive, resulting in different hashes
       sha256 = if cudaSupport then
-        "sha256-ZAi093u2AQBBcgs9O2/KqdoZCtupQGzWc/1BNofEpoI="
+        "sha256-yxEbdEd7c18+N0XQgadjG72fMbD3f6CvMdfejb6kalo="
       else
         if stdenv.isDarwin then
-          "sha256-LLjUcM+s69CbAczuAvytoM3jnq+VnF9XUA6lXC1XJoo="
+          "sha256-SPdW7lBGdlozXBg9nBN1R3XK4stNAgHQO2ewHE2ohCc="
         else
-          "sha256-yOxwOHgk78ezm7/Uo6oHut8r3/xLgnnCBc0mcLNHSRE=";
+          "sha256-zrXOEp2jHf1kYsPca4d1IQS3j6K0tpEoPwld/u4Bn5c=";
     };
 
     buildAttrs = {
@@ -387,7 +389,7 @@ in buildPythonPackage {
     keras-preprocessing
     numpy
     opt-einsum
-    protobuf
+    protobuf-python
     six
     tblib
     tensorflow-estimator_2

--- a/pkgs/development/python-modules/tensorflow/default.nix
+++ b/pkgs/development/python-modules/tensorflow/default.nix
@@ -126,6 +126,9 @@ let
       ./workspace.patch
       # Patch the sources to compile with protobuf >= 3.16.
       ./system-protobuf.patch
+      # Avoid using a deprecated method that was removed
+      # https://github.com/protocolbuffers/protobuf/blob/66dc42d891a4fc8e9190c524fd67961688a37bbe/src/google/protobuf/io/coded_stream.h#L387-L39
+      ./system-protobuf_SetTotalBytesLimit.patch
     ];
 
     # On update, it can be useful to steal the changes from gentoo

--- a/pkgs/development/python-modules/tensorflow/system-protobuf.patch
+++ b/pkgs/development/python-modules/tensorflow/system-protobuf.patch
@@ -1,0 +1,13 @@
+diff --git a/tensorflow/core/kernels/example_parsing_ops.cc b/tensorflow/core/kernels/example_parsing_ops.cc
+index a1265cfb5c6..ada919bbd7b 100644
+--- a/tensorflow/core/kernels/example_parsing_ops.cc
++++ b/tensorflow/core/kernels/example_parsing_ops.cc
+@@ -1218,7 +1218,7 @@ class DecodeJSONExampleOp : public OpKernel {
+           resolver_.get(), "type.googleapis.com/tensorflow.Example", &in, &out);
+       OP_REQUIRES(ctx, status.ok(),
+                   errors::InvalidArgument("Error while parsing JSON: ",
+-                                          string(status.error_message())));
++                                          string(status.message())));
+     }
+   }
+ 

--- a/pkgs/development/python-modules/tensorflow/system-protobuf_SetTotalBytesLimit.patch
+++ b/pkgs/development/python-modules/tensorflow/system-protobuf_SetTotalBytesLimit.patch
@@ -1,0 +1,21 @@
+diff -ru a/tensorflow/core/distributed_runtime/tensor_coding.cc b/tensorflow/core/distributed_runtime/tensor_coding.cc
+--- a/tensorflow/core/distributed_runtime/tensor_coding.cc	2021-12-19 19:12:25.000000000 -0500
++++ b/tensorflow/core/distributed_runtime/tensor_coding.cc	2021-12-20 16:10:32.000000000 -0500
+@@ -82,7 +82,7 @@
+ Status TensorResponse::ParseFrom(Source* source) {
+   if (!on_host_) {
+     protobuf::io::CodedInputStream input(source->contents());
+-    input.SetTotalBytesLimit(INT_MAX, INT_MAX);  // Unlimited
++    input.SetTotalBytesLimit(INT_MAX);  // Unlimited
+ 
+     // Pre-parse into local storage, then delegate to device.
+     if (!meta_.ParseFromCodedStream(&input) || !input.ConsumedEntireMessage()) {
+@@ -218,7 +218,7 @@
+ 
+ bool TensorResponse::ParseFast(Source* source) {
+   protobuf::io::CodedInputStream input(source->contents());
+-  input.SetTotalBytesLimit(INT_MAX, INT_MAX);  // Unlimited
++  input.SetTotalBytesLimit(INT_MAX);  // Unlimited
+   while (true) {
+     auto p = input.ReadTagWithCutoff(127);
+     int tag = GetTagFieldNumber(p.first);

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -9240,6 +9240,8 @@ in {
     inherit (pkgs.darwin.apple_sdk.frameworks) Foundation Security;
     flatbuffers-core = pkgs.flatbuffers;
     flatbuffers-python = self.flatbuffers;
+    protobuf-core = pkgs.protobuf;
+    protobuf-python = self.protobuf;
     lmdb-core = pkgs.lmdb;
   };
 

